geode-core/src/main/java/org/apache/geode/internal/tcp/MsgReader.java

```diff
@@ -14,70 +14,65 @@
  */
 package org.apache.geode.internal.tcp;
 
+import java.io.EOFException;
 import java.io.IOException;
-import java.nio.BufferUnderflowException;
+import java.io.InputStream;
+import java.net.Socket;
 import java.nio.ByteBuffer;
 
-import org.apache.logging.log4j.Logger;
-
 import org.apache.geode.distributed.internal.DMStats;
 import org.apache.geode.distributed.internal.DistributionMessage;
 import org.apache.geode.distributed.internal.ReplyProcessor21;
 import org.apache.geode.internal.Assert;
 import org.apache.geode.internal.InternalDataSerializer;
 import org.apache.geode.internal.net.BufferPool;
-import org.apache.geode.internal.net.NioFilter;
+import org.apache.geode.internal.net.SocketUtils;
 import org.apache.geode.internal.serialization.Version;
-import org.apache.geode.logging.internal.log4j.api.LogService;
 
 /**
  * This class is currently used for reading direct ack responses It should probably be used for all
  * of the reading done in Connection.
  *
  */
 public class MsgReader {
-  private static final Logger logger = LogService.getLogger();
-
-  protected final Connection conn;
+  protected final ClusterConnection conn;
+  private final BufferPool bufferPool;
   protected final Header header = new Header();
-  private final NioFilter ioFilter;
   private ByteBuffer peerNetData;
+  private final InputStream inputStream;
   private final ByteBufferInputStream byteBufferInputStream;
+  private int lastProcessedPosition;
+  private int lastReadPosition;
```

Comment:
Both of those variables are state used in readAtLeast()
```
    // keep track of how much of the buffer contains valid data with lastReadPosition
    lastReadPosition = buffer.position();

    // set up the buffer for reading and keep track of how much has been consumed with
    // lastProcessedPosition
    buffer.limit(lastProcessedPosition + bytes);
    buffer.position(lastProcessedPosition);
    lastProcessedPosition += bytes;
```

Reply:
My comment came out of frustration, perhaps born of laziness. I'm trying to decide whether or not I think the new code in `MsgReader` is correct—by inspection. That's very hard for me.

Earlier in the day I went and read the NIO `Buffer` code since I needed to understand that code to understand these changes. That class has [Jon Bentley](https://www.goodreads.com/book/show/52084.Programming_Pearls)-level comments detailing the purpose of each field along with the field invariants. The documentation of those invariants are a huge aid in visual inspection. 

https://docs.oracle.com/javase/8/docs/api/java/nio/Buffer.html

So with this in mind:

1. my hunch is that `Buffer` already has all the state we need to solve this problem—so I am suspicious of these added fields
2. it feels like we're doing a lot of poking into the `Buffer` object to control its details—setting limit and position explicitly rather than putting and getting and flipping

If the extra fields are really needed then it would help to see a block comment at the point of declaration, tying together these variables with their invariants. Rather than saying things like "keep track of", it would be helpful to know the relationships between these two variables and between them and the various state variables inside the `Buffer`.

Reply:
I'm not going to change that logic.  It's what used to be in the MsgReader subclasses before I moved it to the NioFilter classes.  This PR just restores that logic to MsgReader.

Reply:
Let me say a little more about these variables - my understanding of doing this external book-keeping on the buffer is to avoid a compaction when there is still room in the buffer to read more data.  The code in ClusterConnection that reads messages still does that compaction and we should eventually remove it and use MsgReader to read all messages.

Reply:
thanks for that clarification @bschuchardt. I thought maybe this code was old (and therefore vetted) but it wasn't obvious from the PR.